src/twisted/internet/test/test_tcp.py

```diff
@@ -913,10 +931,367 @@ def createTestSocket(test, addressFamily, socketType):
+        """
+        self.exhauster.exhaust()
+        self.exhauster.release()
+        self.openAFile()        # Does not fail with EMFILE
+
+
+    def test_fileDescriptorsReleasedOnFailure(self):
+        """
+        L{_ExhaustsFileDescriptors.exhaust} closes any opened file
+        descriptors if an exception occurs during its exhaustion loop.
+        """
+        fileDescriptors = []
+        def failsAfterThree():
+            if len(fileDescriptors) == 3:
+                raise ValueError
+            else:
+                fd = os.dup(0)
+                fileDescriptors.append(fd)
+                return fd
+
+        exhauster = _ExhaustsFileDescriptors(failsAfterThree)
+        self.addCleanup(exhauster.release)
+
+        self.assertRaises(ValueError, exhauster.exhaust)
+        self.assertEqual(len(fileDescriptors), 3)
+        self.assertEqual(exhauster.count(), 0)
+
+        for fd in fileDescriptors:
+            exception = self.assertRaises(OSError, os.close, fd)
+            self.assertEqual(exception.errno, errno.EBADF)
+
+
+
+def assertPeerClosedOnEMFILE(
+        testCase,
+        exhauster,
+        reactor,
+        runReactor,
+        listen,
+        connect,
+):
+    """
+    Assert that an L{IListeningPort} immediately closes an accepted
+    peer socket when the number of open file descriptors exceeds the
+    soft resource limit.
+
+    @param testCase: The test case under which to run this assertion.
+    @type testCase: L{trial.unittest.SynchronousTestCase} or
+        L{trial.unittest.TestCase}
+
+    @param exhauster: The file descriptor exhauster.
+    @type exhauster: L{_ExhaustsFileDescriptors}
+
+    @param reactor: The reactor under test.
+
+    @param runReactor: A callable that will synchronously run the
+        provided reactor.
+
+    @param listen: A callback to bind to a port.
+    @type listen: A L{callable} that accepts two arguments: the
+        provided C{reactor}; and a L{ServerFactory}.  It must return
+        an L{IListeningPort} provider.
+
+    @param connect: A callback to connect a client to the listening
+        port.
+    @type connect: A L{callable} that accepts three arguments: the
+        provided C{reactor}; the address returned by
+        L{IListeningPort.getHost}; and a L{ClientFactory}.  Its return
+        value is ignored.
+    """
+    testCase.addCleanup(exhauster.release)
+
+    serverFactory = MyServerFactory()
+    serverConnectionMade = Deferred()
+    serverFactory.protocolConnectionMade = serverConnectionMade
+    serverConnectionCompleted = [False]
+
+    def stopReactorIfServerAccepted(_):
+        reactor.stop()
+        serverConnectionCompleted[0] = True
+
+    serverConnectionMade.addCallback(
+        stopReactorIfServerAccepted)
+
+    port = listen(reactor, serverFactory)
+    listeningHost = port.getHost()
+    clientFactory = MyClientFactory()
+    connect(reactor, listeningHost, clientFactory)
+
+    reactor.callWhenRunning(exhauster.exhaust)
```

Comment:
> Moving this to before startup prevents the reactor from creating its signal delivery pipe ￼

Woops!  Guess there's too much random unspecified stuff going on in the reactor to do much better than what's here, then.  Maybe someday.

> FWIW it appears that reactors don't wait for startup events before performing I/O:

True but it does _run_ them.  So it seemed like this would be okay - because `exhaust` doesn't return a Deferred, it just does its job and then it's done.  So not waiting would be okay.  But given the other problems here, it's moot.